### PR TITLE
Introduce CI workflow logic

### LINKED ISSUE #194: Only run tests if code has changed
Set the CI workflow up such that we only build and test FTorch if files that affect it have been modified.

For example, we shouldn't run the tests if only Markdown files have changed.

### PR DESCRIPTION
**Pull request by jwallwork23**
Closes #194.

In the interests of green computing and avoiding hitting the limit of the free-plan, this PR introduces some logic to only run workflows if relevant changes are made.

* Only run the tests if source files, `CMakeLists.txt`, `requirements.txt`, or data are changed.
* Only run the static analysis workflow if C, C++, Fortran, or Python files have changed.
* Only run the Fypp checks if Fortran files have changed.
* Only build the docs if Markdown or content in `pages/` have changed.

If any of the workflow files are changed then those tests should also be run.

The PR also improves the consistency of formatting across the workflow configuration files.

### CODE REVIEW

**Comment by jwallwork23** [comment]
Some checks:
* My initial push https://github.com/Cambridge-ICCS/FTorch/pull/195/commits/00bee3e6c32bdcdf915675cdb098bb55c929d50a didn't run any tests because no source code or docs have been changed.
* My second push https://github.com/Cambridge-ICCS/FTorch/pull/195/commits/de42e78a4316ce461b7e3f673865535362a23492 added a check for if `fypp_checks.yml` changed. This file has changed in the PR so those tests were run.
* My third push https://github.com/Cambridge-ICCS/FTorch/pull/195/commits/d154d1472a08296aca9a1170515f08ea94a94792 added a check for if `static_analysis.yml` changed, which it has. So both Fypp and linting checks were run.
* (You get the picture) https://github.com/Cambridge-ICCS/FTorch/pull/195/commits/3ba39ca7dc62c0c98184f654cd17e67f59a4b79c added `build_docs.yml` check, so Fypp, static analysis and build docs were run.
* Finally https://github.com/Cambridge-ICCS/FTorch/pull/195/commits/c1f6a0449bd7a98a50b545b6195e3b061398d120 added `test_suite.yml` check, so all four were run.

File: .github/workflows/fypp_checks.yml [hunk 1]
```diff
@@ -1,19 +1,29 @@
...
+  # Triggers the workflow on pushes to open pull requests with Fortran changes
   pull_request:
     types:
-    - synchronize
-    - opened
-    - reopened
-
+      - synchronize
+      - opened
+      - reopened
```
Comment (jwallwork23) [inline .github/workflows/fypp_checks.yml:14]: Was this added for a particular reason? We don't mention it in the other workflows.

File: .github/workflows/test_suite.yml [hunk 2]
```diff
@@ -1,17 +1,36 @@
...
+      - '**requirements.txt'
+      - '**data/*'
 
   # Allows you to run this workflow manually from the Actions tab
   workflow_dispatch:
 
+# Cancel jobs running if new commits are pushed
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
```
Comment (jwallwork23) [inline .github/workflows/test_suite.yml:32]: I also added this, which cancels the tests if you push while they are still running. Happy to take it out if preferred.

File: .github/workflows/fypp_checks.yml (hunk 1 above)
Comment (jatkinson1000) [reply .github/workflows/fypp_checks.yml:14]: Hhhhmmmmm, not sure. I think if you have a better solution then consistency is good.

File: .github/workflows/test_suite.yml (hunk 2 above)
Comment (jatkinson1000) [reply .github/workflows/test_suite.yml:32]: Nah, I like this, it is rare, but there have been a couple of times where I would have used this.

**Review by jatkinson1000 — COMMENTED** [review]
This all seems sensible @jwallwork23 
A couple of stylistic comments and a suggestion to remove the options you had already flagged.

File: .github/workflows/fypp_checks.yml (hunk 1 above)
Comment (jwallwork23) [reply .github/workflows/fypp_checks.yml]: Addressed in https://github.com/Cambridge-ICCS/FTorch/pull/195/commits/03a836dc35f7aae8a786dfe94f8086448b36b138.

**Pull request merged**